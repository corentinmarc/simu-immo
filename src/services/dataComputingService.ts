import { range } from 'lodash-es';
import { ComputedData, MonthData } from 'types/data';

const computeCapitalPercent = (
  i: number,
  {
    duration,
    interestRate,
  }: {
    duration: number;
    interestRate: number;
  },
) => (1 / (1 + interestRate / 12 / 100) ** (duration * 12 - i)) * 100;

const getAverageCapitalPercent = ({
  duration,
  interestRate,
}: {
  duration: number;
  interestRate: number;
}) => {
  const nbMonth = Number(duration) * 12 || 0;
  return (
    range(0, nbMonth)
      .map(i => computeCapitalPercent(i, { duration, interestRate }), [])
      .reduce((acc, cur) => acc + cur, 0) / nbMonth
  );
};

// @todo use a worker to make those computations
const computeDatas = ({
  capital,
  duration,
  interestRate,
  insuranceRate,
  notaryRate,
  intercalaryFees,
}: {
  capital: number;
  duration: number;
  interestRate: number;
  insuranceRate: number;
  notaryRate: number;
  intercalaryFees: number;
}): ComputedData => {
  const nbMonth = Number(duration) * 12 || 0;
  const totalInsurancePercent = insuranceRate * duration;
  const totalCapitalPercent = getAverageCapitalPercent({
    duration,
    interestRate,
  });

  let monthsDatas = range(0, nbMonth).map(i => {
    const percentCapital = computeCapitalPercent(i, {
      duration,
      interestRate,
    });
    return {
      percentCapital,
      month: i,
      percentInterest: 100 - percentCapital,
      percentInsurance: totalInsurancePercent,
      ratioCapital:
        percentCapital /
        (100 + (totalInsurancePercent * totalCapitalPercent) / 100),
      ratioInterest:
        (100 - percentCapital) /
        (100 + (totalInsurancePercent * totalCapitalPercent) / 100),
      ratioInsurance:
        (totalInsurancePercent * totalCapitalPercent) /
        100 /
        (100 + (totalInsurancePercent * totalCapitalPercent) / 100),
    };
  }) as MonthData[];

  const totalRatioInterest =
    monthsDatas.reduce((acc, { ratioInterest }) => acc + ratioInterest, 0) /
    nbMonth;
  const totalRatioCapital =
    monthsDatas.reduce((acc, { ratioCapital }) => acc + ratioCapital, 0) /
    nbMonth;
  const totalRatioInsurance =
    monthsDatas.reduce((acc, { ratioInsurance }) => acc + ratioInsurance, 0) /
    nbMonth;
  const totalCost = capital / totalRatioCapital;
  const totalInterestCost = totalCost * totalRatioInterest;
  const totalInsuranceCost = totalCost * totalRatioInsurance;
  const totalCapitalCost = totalCost * totalRatioCapital;
  const mensualityCost = totalCost / duration / 12;
  const notaryCost = (capital * notaryRate) / 100;
  const startCost = notaryCost + Number(intercalaryFees);
  const greatTotalCost = totalCost + startCost;

  monthsDatas = monthsDatas.map((monthData, i) => ({
    ...monthData,
    costInterest: mensualityCost * monthData.ratioInterest,
    costCapital: mensualityCost * monthData.ratioCapital,
    costInsurance: mensualityCost * monthData.ratioInsurance,
    costCumulate: mensualityCost * (i + 1),
  }));

  return {
    totalInterestCost,
    totalInsuranceCost,
    totalCapitalCost,
    greatTotalCost,
    totalRatioCapital,
    totalRatioInterest,
    totalRatioInsurance,
    totalCost,
    mensualityCost,
    notaryCost,
    startCost,
    monthsDatas,
    intercalaryFees: +intercalaryFees,
  };
};

export { computeDatas };
