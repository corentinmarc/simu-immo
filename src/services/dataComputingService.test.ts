import { computeDatas } from './dataComputingService';

describe('Services/dataComputingService', () => {
  describe('computeDatas', () => {
    const dataTest1 = {
      capital: 100000,
      duration: 1,
      interestRate: 2,
      insuranceRate: 0.3,
      notaryRate: 3,
      intercalaryFees: 6000,
    };
    test('Snapshot of computedDatas', () => {
      const cumputedDatas = computeDatas(dataTest1);

      expect(cumputedDatas).toMatchSnapshot();
    });
  });
});
