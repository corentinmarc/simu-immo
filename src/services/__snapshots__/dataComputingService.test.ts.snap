// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Services/dataComputingService computeDatas Snapshot of computedDatas 1`] = `
Object {
  "greatTotalCost": 110386.64074092657,
  "intercalaryFees": 6000,
  "mensualityCost": 8448.886728410547,
  "monthsDatas": Array [
    Object {
      "costCapital": 8257.220061743874,
      "costCumulate": 8448.886728410547,
      "costInsurance": 25.000000000000004,
      "costInterest": 166.66666666667177,
      "month": 0,
      "percentCapital": 98.02149919579796,
      "percentInsurance": 0.3,
      "percentInterest": 1.9785008042020422,
      "ratioCapital": 0.9773145654773466,
      "ratioInsurance": 0.0029589697203460017,
      "ratioInterest": 0.019726464802307282,
    },
    Object {
      "costCapital": 8270.982095180116,
      "costCumulate": 16897.773456821094,
      "costInsurance": 25.000000000000004,
      "costInterest": 152.90463323043198,
      "month": 1,
      "percentCapital": 98.18486836112429,
      "percentInsurance": 0.3,
      "percentInterest": 1.8151316388757124,
      "ratioCapital": 0.9789434230864756,
      "ratioInsurance": 0.0029589697203460017,
      "ratioInterest": 0.01809760719317837,
    },
    Object {
      "costCapital": 8284.767065338749,
      "costCumulate": 25346.66018523164,
      "costInsurance": 25.000000000000004,
      "costInterest": 139.1196630717992,
      "month": 2,
      "percentCapital": 98.34850980839282,
      "percentInsurance": 0.3,
      "percentInterest": 1.6514901916071807,
      "ratioCapital": 0.9805749954582863,
      "ratioInsurance": 0.0029589697203460017,
      "ratioInterest": 0.016466034821367666,
    },
    Object {
      "costCapital": 8298.575010447648,
      "costCumulate": 33795.54691364219,
      "costInsurance": 25.000000000000004,
      "costInterest": 125.31171796289905,
      "month": 3,
      "percentCapital": 98.51242399140683,
      "percentInsurance": 0.3,
      "percentInterest": 1.487576008593166,
      "ratioCapital": 0.9822092871173838,
      "ratioInsurance": 0.0029589697203460017,
      "ratioInterest": 0.014831743162270256,
    },
    Object {
      "costCapital": 8312.405968798394,
      "costCumulate": 42244.43364205274,
      "costInsurance": 25.000000000000004,
      "costInterest": 111.48075961215432,
      "month": 4,
      "percentCapital": 98.67661136472583,
      "percentInsurance": 0.3,
      "percentInterest": 1.3233886352741706,
      "ratioCapital": 0.9838463025959125,
      "ratioInsurance": 0.0029589697203460017,
      "ratioInterest": 0.013194727683741444,
    },
    Object {
      "costCapital": 8326.259978746391,
      "costCumulate": 50693.32037046328,
      "costInsurance": 25.000000000000004,
      "costInterest": 97.62674966415453,
      "month": 5,
      "percentCapital": 98.84107238366707,
      "percentInsurance": 0.3,
      "percentInterest": 1.1589276163329316,
      "ratioCapital": 0.9854860464335726,
      "ratioInsurance": 0.0029589697203460017,
      "ratioInterest": 0.011554983846081297,
    },
    Object {
      "costCapital": 8340.13707871097,
      "costCumulate": 59142.20709887383,
      "costInsurance": 25.000000000000004,
      "costInterest": 83.74964969957689,
      "month": 6,
      "percentCapital": 99.00580750430652,
      "percentInsurance": 0.3,
      "percentInterest": 0.9941924956934827,
      "ratioCapital": 0.9871285231776287,
      "ratioInsurance": 0.0029589697203460017,
      "ratioInterest": 0.009912507102025304,
    },
    Object {
      "costCapital": 8354.03730717549,
      "costCumulate": 67591.09382728438,
      "costInsurance": 25.000000000000004,
      "costInterest": 69.84942123505819,
      "month": 7,
      "percentCapital": 99.17081718348037,
      "percentInsurance": 0.3,
      "percentInterest": 0.8291828165196335,
      "ratioCapital": 0.9887737373829247,
      "ratioInsurance": 0.0029589697203460017,
      "ratioInterest": 0.008267292896729207,
    },
    Object {
      "costCapital": 8367.960702687447,
      "costCumulate": 76039.98055569493,
      "costInsurance": 25.000000000000004,
      "costInterest": 55.926025723099066,
      "month": 8,
      "percentCapital": 99.33610187878617,
      "percentInsurance": 0.3,
      "percentInterest": 0.6638981212138333,
      "ratioCapital": 0.9904216936118962,
      "ratioInsurance": 0.0029589697203460017,
      "ratioInterest": 0.0066193366677576694,
    },
    Object {
      "costCapital": 8381.907303858594,
      "costCumulate": 84488.86728410548,
      "costInsurance": 25.000000000000004,
      "costInterest": 41.97942455195267,
      "month": 9,
      "percentCapital": 99.50166204858415,
      "percentInsurance": 0.3,
      "percentInterest": 0.4983379514158486,
      "ratioCapital": 0.9920723964345829,
      "ratioInsurance": 0.0029589697203460017,
      "ratioInterest": 0.004968633845071098,
    },
    Object {
      "costCapital": 8395.877149365026,
      "costCumulate": 92937.75401251602,
      "costInsurance": 25.000000000000004,
      "costInterest": 28.009579045521672,
      "month": 10,
      "percentCapital": 99.66749815199846,
      "percentInsurance": 0.3,
      "percentInterest": 0.3325018480015416,
      "ratioCapital": 0.9937258504286405,
      "ratioInsurance": 0.0029589697203460017,
      "ratioInterest": 0.0033151798510134595,
    },
    Object {
      "costCapital": 8409.8702779473,
      "costCumulate": 101386.64074092657,
      "costInsurance": 25.000000000000004,
      "costInterest": 14.01645046324579,
      "month": 11,
      "percentCapital": 99.83361064891847,
      "percentInsurance": 0.3,
      "percentInterest": 0.1663893510815342,
      "ratioCapital": 0.995382060179355,
      "ratioInsurance": 0.0029589697203460017,
      "ratioInterest": 0.001658970100298959,
    },
  ],
  "notaryCost": 3000,
  "startCost": 9000,
  "totalCapitalCost": 100000,
  "totalCost": 101386.64074092657,
  "totalInsuranceCost": 300,
  "totalInterestCost": 1086.640740926565,
  "totalRatioCapital": 0.9863232401153338,
  "totalRatioInsurance": 0.0029589697203460017,
  "totalRatioInterest": 0.010717790164320166,
}
`;
